library/VTKExtensions/Rendering/vtkF3DHexagonalBokehBlurPass.cxx

```diff
@@ -0,0 +1,349 @@
+    this->DiagonalBlurTexture->SetWrapS(vtkTextureObject::ClampToEdge);
+    this->DiagonalBlurTexture->SetWrapT(vtkTextureObject::ClampToEdge);
+    this->DiagonalBlurTexture->Allocate2D(w, h, 4, VTK_FLOAT);
+  }
+
+  if (this->FrameBufferObject == nullptr)
+  {
+    this->FrameBufferObject = vtkSmartPointer<vtkOpenGLFramebufferObject>::New();
+    this->FrameBufferObject->SetContext(renWin);
+  }
+}
+
+//------------------------------------------------------------------------------
+void vtkF3DHexagonalBokehBlurPass::RenderDelegate(const vtkRenderState* s, int w, int h)
+{
+  this->PreRender(s);
+
+  this->FrameBufferObject->GetContext()->GetState()->PushFramebufferBindings();
+  this->FrameBufferObject->Bind();
+
+  this->FrameBufferObject->AddColorAttachment(0, this->BackgroundTexture);
+  this->FrameBufferObject->ActivateDrawBuffers(1);
+  this->FrameBufferObject->StartNonOrtho(w, h);
+
+  vtkOpenGLRenderer* glRen = vtkOpenGLRenderer::SafeDownCast(s->GetRenderer());
+
+  glRen->GetState()->vtkglClear(GL_COLOR_BUFFER_BIT);
+
+  this->DelegatePass->Render(s);
+  this->NumberOfRenderedProps += this->DelegatePass->GetNumberOfRenderedProps();
+
+  this->FrameBufferObject->RemoveColorAttachments(1);
+
+  this->FrameBufferObject->GetContext()->GetState()->PopFramebufferBindings();
+
+  this->PostRender(s);
+}
+
+//------------------------------------------------------------------------------
+void vtkF3DHexagonalBokehBlurPass::RenderDirectionalBlur(
+  vtkOpenGLRenderWindow* renWin, int w, int h)
+{
+  if (!this->BlurQuadHelper)
+  {
+    std::string FSSource = vtkOpenGLRenderUtilities::GetFullScreenQuadFragmentShaderTemplate();
+
+    std::stringstream ssDecl;
+    ssDecl << "uniform sampler2D backgroundTexture;\n";
+    ssDecl << "uniform vec2 invViewDims;\n";
+    ssDecl << "uniform float coc;\n";
+    ssDecl << BlurFunc();
+    ssDecl << "//VTK::FSQ::Decl";
+
+    vtkShaderProgram::Substitute(FSSource, "//VTK::FSQ::Decl", ssDecl.str());
+
+    std::stringstream ssImpl;
+
+    ssImpl << "  vec2 blurDir = vec2(cos(PI/2), sin(PI/2));\n";
+    ssImpl << "  vec3 color1 = BlurTexture(backgroundTexture, texCoord, blurDir).rgb;\n";
+    ssImpl << "  blurDir = vec2(cos(-PI/6), sin(-PI/6));\n";
+    ssImpl << "  vec3 color2 = BlurTexture(backgroundTexture, texCoord, blurDir).rgb;\n";
+    ssImpl << "  gl_FragData[0] = vec4(color1, 1.0);\n";
+    ssImpl << "  gl_FragData[1] = vec4(color1 + color2, 1.0);\n";
+
+    vtkShaderProgram::Substitute(FSSource, "//VTK::FSQ::Impl", ssImpl.str());
+
+    this->BlurQuadHelper = new vtkOpenGLQuadHelper(renWin,
+      vtkOpenGLRenderUtilities::GetFullScreenQuadVertexShader().c_str(), FSSource.c_str(), "");
+  }
+
+  renWin->GetShaderCache()->ReadyShaderProgram(this->BlurQuadHelper->Program);
+
+  if (!this->BlurQuadHelper->Program || !this->BlurQuadHelper->Program->GetCompiled())
+  {
+    vtkErrorMacro("Couldn't build the Blur shader program.");
+    return;
+  }
```

Comment:
can this actually happen ?

Reply:
Who knows? It depends on the driver. It's better to gracefully exit than crashing.

Reply:
ok. not testable I suppose ?

Reply:
nop